{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8849394",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import shlex\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import fnmatch\n",
    "import plotly.io as pio\n",
    "import scipy\n",
    "from scipy.signal import find_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ee4e169",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('noknownsnps.table', sep=\"\\t\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1c9a89d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def smAlleleFreq(sizeOfBulk, rep):\n",
    "#     '''\n",
    "#     An AA/Aa/aa individual carries 0%, 50%, and 100% of the alt (a) allele, respectively.\n",
    "#     The AA:Aa:aa ratios are 0.25:0.5:0.25 in a F2 population (null hypothesis).\n",
    "#     '''\n",
    "    \n",
    "#     freqL = []\n",
    "#     pop = [0.0, 0.5, 1.0]\n",
    "    \n",
    "#     prob = [0.25, 0.5, 0.25]\n",
    "\n",
    "#     altFreq = np.random.choice(pop, sizeOfBulk, p=prob).mean()\n",
    "#     freqL.append(altFreq)\n",
    "\n",
    "#     return sum(freqL)/len(freqL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa04658e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gStatistic_Array(o1, o3, o2, o4):\n",
    "    # Calculate G-statistc using numpy array input\n",
    "    np.seterr(all='ignore')\n",
    "\n",
    "    e1 = np.where(o1+o2+o3+o4!=0, (o1+o2)*(o1+o3)/(o1+o2+o3+o4), 0)\n",
    "    e2 = np.where(o1+o2+o3+o4!=0, (o1+o2)*(o2+o4)/(o1+o2+o3+o4), 0)\n",
    "    e3 = np.where(o1+o2+o3+o4!=0, (o3+o4)*(o1+o3)/(o1+o2+o3+o4), 0)\n",
    "    e4 = np.where(o1+o2+o3+o4!=0, (o3+o4)*(o2+o4)/(o1+o2+o3+o4), 0)\n",
    "\n",
    "    llr1 = np.where(o1/e1>0, 2*o1*np.log(o1/e1), 0.0)\n",
    "    llr2 = np.where(o2/e2>0, 2*o2*np.log(o2/e2), 0.0)\n",
    "    llr3 = np.where(o3/e3>0, 2*o3*np.log(o3/e3), 0.0)\n",
    "    llr4 = np.where(o4/e4>0, 2*o4*np.log(o4/e4), 0.0)\n",
    "\n",
    "    return np.where(e1*e2*e3*e4==0, 0.0, llr1+llr2+llr3+llr4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60f853b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate delta SNP ratio\n",
    "df['ratio'] = ((df.wt_ref)/(df.wt_ref+df.wt_alt))-((df.mu_ref)/(df.mu_ref+df.mu_alt))\n",
    "\n",
    "#Drop indels, leave only SNPs\n",
    "df[\n",
    "    df[\"ref\"].apply(lambda x: len(x) == 1)\n",
    "    & df[\"alt\"].apply(lambda x: len(x) == 1)\n",
    "]\n",
    "\n",
    "#Calculate ratio_mean for peak identification lower limit\n",
    "ratio_mean = df[\"ratio\"].mean()\n",
    "\n",
    "#df.drop(df[df['ratio'] <= 0.1].index, inplace = True) # drop ratios below the defined value. Sometimes necissary\n",
    "\n",
    "#Drop NAs. Infrequently will arise while div/0 seemingly\n",
    "df.dropna(axis=0, how='any', subset=\"ratio\", inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6339945",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#lowess smoothing by chromosome\n",
    "chr_facets=df[\"chr\"].unique()\n",
    "df_list = []\n",
    "\n",
    "lowess = sm.nonparametric.lowess\n",
    "lowess_span=0.3\n",
    "\n",
    "for i in chr_facets:\n",
    "    \n",
    "    # Fit LOWESS ratio ~ position\n",
    "    df_chr = df[df['chr']==i].copy()\n",
    "\n",
    "    X=df_chr['pos'].values\n",
    "    \n",
    "    Y=df_chr['ratio'].values\n",
    "    \n",
    "    y_hat = lowess(Y,X, frac=lowess_span)[:,1]\n",
    "    \n",
    "    df_chr['yhat'] = y_hat # fitted values\n",
    "    \n",
    "    df_list.append(df_chr)\n",
    "    \n",
    "    # Find peaks in the LOWESS fitted curve\n",
    "    signal = df_chr['yhat'].to_numpy().flatten()\n",
    "    \n",
    "    peaks = scipy.signal.find_peaks(\n",
    "        signal, \n",
    "        height=0.5, \n",
    "        threshold=None, \n",
    "        distance=None, \n",
    "        prominence=None, \n",
    "        width=None,\n",
    "        wlen=None, \n",
    "        rel_height=None,\n",
    "        plateau_size=None)\n",
    "    \n",
    "    if len(peaks[0]) > 0:\n",
    "    \n",
    "        pi = peaks[0] \n",
    "        h = peaks[1]['peak_heights']\n",
    "\n",
    "        # Get index of the tallest peak\n",
    "        pi_max = (pi[np.argmax(h)]).flatten()\n",
    "\n",
    "        peak_widths = scipy.signal.peak_widths(\n",
    "            signal, \n",
    "            peaks=pi_max, \n",
    "            rel_height=0.5, \n",
    "            prominence_data=None, \n",
    "            wlen=None)\n",
    "\n",
    "        peak_bounds = peak_widths[2:4]\n",
    "\n",
    "        peak_bound_min = np.round(peak_bounds[0])\n",
    "        peak_bound_max = np.round(peak_bounds[-1])\n",
    "\n",
    "        peak_bound_min_pos = df_chr['pos'].iloc[peak_bound_min].item()\n",
    "        peak_bound_max_pos = df_chr['pos'].iloc[peak_bound_max].item()\n",
    "\n",
    "    # label peaks\n",
    "        if len(peak_bounds[0]) != 0:\n",
    "            df_chr['peak'] = [1 if x >= peak_bound_min_pos and x <= peak_bound_max_pos else 0 for x in df_chr['pos']]\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        df_chr['peak'] = 0\n",
    "        \n",
    "df = pd.concat(df_list)\n",
    "\n",
    "df_peaks = df.loc[df['peak'] == 1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c5c846",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate G-statistic to glean more info about potential causal SNPs\n",
    "df['G_S'] = gStatistic_Array(df['wt_ref'], df['wt_alt'], df['mu_ref'], df['mu_alt'])\n",
    "\n",
    "#calculate 98th percentile of G-statistics\n",
    "GS_98p = np.percentile(df['G_S'], 90)\n",
    "\n",
    "df['G_S_98p'] = [1 if (np.isclose(x, GS_98p) or (x > GS_98p))\n",
    "                else 0 for x in df['G_S']]\n",
    "\n",
    "dfLC11=df.loc[(df['peak'] == 1) & (df['G_S_98p'] == 1)]\n",
    "dfLC01=df.loc[(df['peak'] == 0) & (df['G_S_98p'] == 1)]\n",
    "dfLC10=df.loc[(df['peak'] == 1) & (df['G_S_98p'] == 0)]\n",
    "\n",
    "df_likely_cands = pd.concat([dfLC11, dfLC01, dfLC10])\n",
    "\n",
    "df_likely_cands.to_csv('likely_candidates.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87002b06",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "chr_facets_p=df_peaks[\"chr\"].unique()\n",
    "\n",
    "#establish fig\n",
    "fig = px.scatter(\n",
    "    df,\n",
    "    x=df['pos'],\n",
    "    y=df['ratio'],\n",
    "    facet_col=\"chr\",\n",
    "    opacity=0.8,\n",
    "    color_discrete_sequence=['goldenrod'],\n",
    "    trendline=\"lowess\",\n",
    "    trendline_options=dict(frac=lowess_span),\n",
    "    trendline_color_override=\"blue\",\n",
    ")\n",
    "            \n",
    "#format fig\n",
    "fig.update_traces(\n",
    "    marker=dict(size=3)\n",
    ")\n",
    "\n",
    "fig.update_layout(\n",
    "    dict(plot_bgcolor = 'white')\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    matches=None\n",
    ")\n",
    "\n",
    "fig.for_each_xaxis(\n",
    "    lambda xaxis: xaxis.update(showticklabels=True)\n",
    ")\n",
    "\n",
    "fig.for_each_xaxis(\n",
    "    lambda xaxis: fig.update_xaxes(tickangle=0)\n",
    ")\n",
    "\n",
    "fig.for_each_xaxis(\n",
    "    lambda x: x.update(title = '')\n",
    ")\n",
    "\n",
    "fig.for_each_yaxis(\n",
    "    lambda y: y.update(title = '')\n",
    ")\n",
    "\n",
    "\n",
    "fig.add_annotation(\n",
    "    showarrow=False,\n",
    "    xanchor='center',\n",
    "    xref='paper', \n",
    "    x=0.5, \n",
    "    yref='paper',\n",
    "    y=-0.1,\n",
    "    text='Position'\n",
    ")\n",
    "\n",
    "fig.add_annotation(\n",
    "    showarrow=False,\n",
    "    xanchor='center',\n",
    "    xref='paper', \n",
    "    x=-0.8, \n",
    "    yanchor='middle',\n",
    "    yref='paper',\n",
    "    y=0.6,\n",
    "    textangle=-90,\n",
    "    text='Ratio'\n",
    ")\n",
    "\n",
    "\n",
    "fig.update_xaxes(\n",
    "    showgrid=True, \n",
    "    gridwidth=0.5, \n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    showline=True, \n",
    "    linewidth=1, \n",
    "    linecolor='black'\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    rangemode=\"tozero\"\n",
    ")\n",
    "\n",
    "fig.update_yaxes(\n",
    "    showgrid=True, \n",
    "    gridwidth=0.5, \n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "\n",
    "fig.update_yaxes(\n",
    "    showline=True, \n",
    "    linewidth=1, \n",
    "    linecolor='black'\n",
    ")\n",
    "\n",
    "fig.update_yaxes(\n",
    "    range=[0, 1]\n",
    ")\n",
    "\n",
    "fig.update_layout(\n",
    "    title=dict(\n",
    "        text=\"Linkage map\",font=dict(color='black')\n",
    "    )\n",
    ")\n",
    "\n",
    "\n",
    "fig['data'][0]['showlegend'] = True\n",
    "fig['data'][0]['name'] = 'Polymorphism ratio'\n",
    "\n",
    "max_pos=df_peaks['pos'].max()\n",
    "min_pos=df_peaks['pos'].min()\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Scatter(\n",
    "        x=[2],\n",
    "        y=[2],\n",
    "        mode='lines',\n",
    "        name='Lowess Fitted Ratio',\n",
    "        line=dict(color=\"blue\")))\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Scatter(\n",
    "        x=[0,0], \n",
    "        y=[0,0], \n",
    "        mode='lines', \n",
    "        line=dict(color='black', width=1, dash='dot'),\n",
    "        name='Identified Peak',\n",
    "    )\n",
    ")\n",
    "\n",
    "#add g-stat > 98 percentile markers\n",
    "for i in chr_facets:\n",
    "    df_chr = df[df['chr']==i].copy()\n",
    "    outliers = df_chr.loc[df_chr['G_S_98p'] == 1 , [\"gene\",\"pos\",\"ratio\"]]\n",
    "    fig.add_trace(\n",
    "    go.Scatter(\n",
    "        x=outliers['pos'],\n",
    "        y=outliers['ratio'],\n",
    "        mode=\"markers\",\n",
    "        name=\"G-stat > 98 percentile\",\n",
    "        line=go.scatter.Line(color=\"red\"),\n",
    "        marker=dict(size=4),\n",
    "        showlegend=False), row=1, col=i\n",
    "    )\n",
    "    \n",
    "#add peak markers\n",
    "for i in chr_facets_p:\n",
    "    df_peaks_chr=df_peaks[df_peaks['chr']==i]\n",
    "    df_chr=df[df['chr'] == i]\n",
    "    max_pos=df_peaks_chr['pos'].max()\n",
    "    min_pos=df_peaks_chr['pos'].min()\n",
    "\n",
    "    if max_pos > 0:\n",
    "        max_yhat_pos=df_peaks_chr.loc[df_peaks_chr['yhat'] == df_peaks_chr['yhat'].max()]['pos'].values[0]\n",
    "        fig.add_vrect(\n",
    "            x0=min_pos, \n",
    "            x1=max_pos, \n",
    "            col=i,\n",
    "            fillcolor=\"red\", \n",
    "            opacity=0.2, \n",
    "            line_width=0\n",
    "        )\n",
    "        \n",
    "        fig.add_vline(\n",
    "            x=max_yhat_pos, \n",
    "            line_dash=\"dot\", \n",
    "            col=i, \n",
    "            line_width=1\n",
    "       )\n",
    "\n",
    "fig.show()\n",
    "\n",
    "#pio.write_image(fig, \"./plotname.png\", scale=6, width=1080, height=540)\n",
    "#df_peaks.to_csv('peaks_file_name.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907df3e8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df['T_S'] = df['G_S']*df['ratio']\n",
    "\n",
    "#establish g-stat lowess fig\n",
    "fig = px.scatter(df, x=df['pos'], y=df['T_S'],\n",
    "    facet_col=\"chr\",\n",
    "    opacity=0.8,\n",
    "    color_discrete_sequence=['goldenrod'],\n",
    "    trendline=\"lowess\",\n",
    "    trendline_options=dict(frac=0.3),\n",
    "    trendline_color_override=\"blue\")\n",
    "\n",
    "#format g-stat lowess fig\n",
    "fig.update_traces(\n",
    "    marker=dict(size=3)\n",
    ")\n",
    "\n",
    "fig.update_layout(\n",
    "    dict(plot_bgcolor = 'white')\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    matches=None\n",
    ")\n",
    "\n",
    "fig.for_each_xaxis(\n",
    "    lambda xaxis: xaxis.update(showticklabels=True)\n",
    ")\n",
    "\n",
    "fig.for_each_xaxis(\n",
    "    lambda xaxis: fig.update_xaxes(tickangle=0)\n",
    ")\n",
    "\n",
    "fig.for_each_xaxis(\n",
    "    lambda x: x.update(title = '')\n",
    ")\n",
    "\n",
    "fig.for_each_yaxis(\n",
    "    lambda y: y.update(title = '')\n",
    ")\n",
    "\n",
    "\n",
    "fig.add_annotation(\n",
    "    showarrow=False,\n",
    "    xanchor='center',\n",
    "    xref='paper', \n",
    "    x=0.5, \n",
    "    yref='paper',\n",
    "    y=-0.1,\n",
    "    text='Position'\n",
    ")\n",
    "\n",
    "fig.add_annotation(\n",
    "    showarrow=False,\n",
    "    xanchor='center',\n",
    "    xref='paper', \n",
    "    x=-0.8, \n",
    "    yanchor='middle',\n",
    "    yref='paper',\n",
    "    y=0.6,\n",
    "    textangle=-90,\n",
    "    text='Ratio'\n",
    ")\n",
    "\n",
    "\n",
    "fig.update_xaxes(\n",
    "    showgrid=True, \n",
    "    gridwidth=0.5, \n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    showline=True, \n",
    "    linewidth=1, \n",
    "    linecolor='black'\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    rangemode=\"tozero\"\n",
    ")\n",
    "\n",
    "fig.update_yaxes(\n",
    "    showgrid=True, \n",
    "    gridwidth=0.5, \n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "\n",
    "fig.update_yaxes(\n",
    "    showline=True, \n",
    "    linewidth=1, \n",
    "    linecolor='black'\n",
    ")\n",
    "\n",
    "fig.update_layout(\n",
    "    title=dict(\n",
    "        text=\"Linkage map\",font=dict(color='black')\n",
    "    )\n",
    ")\n",
    "\n",
    "\n",
    "fig.add_annotation(\n",
    "    showarrow=False,\n",
    "    xanchor='center',\n",
    "    xref='paper', \n",
    "    x=0.5, \n",
    "    yref='paper',\n",
    "    y=-0.1,\n",
    "    text='Position'\n",
    ")\n",
    "\n",
    "fig.add_annotation(\n",
    "    showarrow=False,\n",
    "    xanchor='center',\n",
    "    xref='paper', \n",
    "    x=-0.8, \n",
    "    yanchor='middle',\n",
    "    yref='paper',\n",
    "    y=0.6,\n",
    "    textangle=-90,\n",
    "    text='Ratio'\n",
    ")\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Scatter(\n",
    "    x=[1],\n",
    "    y=[1],\n",
    "    mode='lines',\n",
    "    name='Lowess Fitted G-stat',\n",
    "    line=dict(color=\"blue\")\n",
    "    )\n",
    ")\n",
    "\n",
    "\n",
    "fig['data'][0]['showlegend'] = True\n",
    "fig['data'][0]['name'] = 'G-statistic'\n",
    "\n",
    "max_pos=df_peaks['pos'].max()\n",
    "min_pos=df_peaks['pos'].min()\n",
    "\n",
    "    \n",
    "fig.add_trace(go.Scatter(x=[0,0], \n",
    "    y=[0,0], \n",
    "    mode='lines', \n",
    "    line=dict(color='black', width=1, dash='dot'),\n",
    "    name='Identified Peak',))\n",
    "\n",
    "\n",
    "pio.write_image(fig, \"./plotname.png\", scale=6, width=1080, height=540)\n",
    "df_peaks.to_csv('peaks_file_name.tsv', sep='\\t')\n",
    "\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e7cd743",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2af75d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "327313ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "053d201f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-pyatbsa]",
   "language": "python",
   "name": "conda-env-.conda-pyatbsa-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
