# PyAtBSA
This is my attempt to write a python script which can be run on linux or mac machines using conda to analyze and visualize bulk segrigant analysis (BSA) data generated using illumina sequencing. 

Conceptually, there is nothing new here. My hope is to consolidate multiple techniques for identifiying QTLs or SNPs responsible for a Arabidopsis Thaliana phenotypes using BSA. 

Experimental design is fairly simple. You take an easy to identify phenotype generated through EMS mutageneses, and backcross that mutant into a parental line (Col-0, WS-0, LER, exc) and generate an F2 segrigating population. 

Pick roughly equal amounts of tissue (either though whole leaves or leaf disks) from multiple individuals in each population (12 minimum), extract the pooled genomic DNA and sequence. There are a number of papers on this subject:






